src/main.cpp

```diff
@@ -403,39 +403,318 @@ bool FullSyncWithDPORNodes()
             return true;
 }
 
-double GetPoSKernelPS()
+double GetPoSKernelPS(int nPoSInterval)
 {
-    int nPoSInterval = 72;
-    double dStakeKernelsTriedAvg = 0;
-    int nStakesHandled = 0, nStakesTime = 0;
+    // The number of stakes to include in the average has been reduced to 40(default) from 72. 72 stakes represented 1.8 hours at
+    // standard spacing. This is too long. 40 blocks is nominally 1 hour.
+    double dStakeableWeightSum = 0;
+    double dDiff = 1.0;
+    // Initialize nStakesHandled with -1 because the first iteration of the while loop below is degenerate.
+    // It has to go through once to populate both the previous and current stake in the loop to calculate the staking time.
+    int nStakesHandled = -1;
+    int nStakesTime = 0;
+    double result = 0;
 
-    CBlockIndex* pindex = pindexBest;;
+    CBlockIndex* pindex = pindexBest;
     CBlockIndex* pindexPrevStake = NULL;
 
     while (pindex && nStakesHandled < nPoSInterval)
     {
         if (pindex->IsProofOfStake())
         {
-            dStakeKernelsTriedAvg += GetDifficulty(pindex) * 4294967296.0;
+            /*
+             * An estimate for the stakeable balance * 80 (weight units) across the entire network is
+             * netweight = Diff * (MaxHash / StandardDifficultyTarget) * (16 sec) / TimeBetweenStakes
+             * The first two terms compute the inverse probability of one bernoulli toss (stake test)
+             * The 16 comes from the quantum time unit imposed by the time mask. The TimeBetweenStakes
+             * should nominally be 90 seconds, but in fact varies. To avoid near infinities if TimeBetweenStakes
+             * happens to be very low due to random variation, we will use sum(WeightSum)/sum(StakesTime)
+             * approach to compute the average.
+             *
+             * Don't get confused here. The loop is going backwards in time. Therefore pIndexPrevStake is a later block
+             * here than pindex.
+             */
             nStakesTime += pindexPrevStake ? (pindexPrevStake->nTime - pindex->nTime) : 0;
+            if (fDebug10) LogPrintf("nStakesTime = %u", nStakesTime);
+            dDiff = GetDifficulty(pindex);
+            dStakeableWeightSum += dDiff * 68720525328.0;
+            if (fDebug10) LogPrintf("dDiff = %f", dDiff);
+            if (fDebug10) LogPrintf("dStakeableWeightSum = %f", dStakeableWeightSum);
             pindexPrevStake = pindex;
             nStakesHandled++;
+            if (fDebug10) LogPrintf("nStakesHandled = %u", nStakesHandled);
         }
 
         pindex = pindex->pprev;
     }
 
-    double result = 0;
+    result = nStakesHandled ? dStakeableWeightSum / (double) nStakesTime : 0;
```

Comment:
You are right I think. I will subtract one from the counter at the end, because the first trip through doesn’t actually do anything.

Reply:
It will still be not zero if there are no stakes. If you make the argument unsigned and the counter start at 0 you'll be safe.

Reply:
That is what I meant. I will start the counter at zero, but the final value will be one greater than the actual stakes handled...

Reply:
That is what I meant. I will start the counter at zero, but the final value will be one greater than the actual stakes handled...

Reply:
That is what I meant. I will start the counter at zero, but the final value will be one greater than the actual stakes handled...

Reply:
Ok. @denravonska through Slack, before the comments above, suggested I rename GetPoSKernelPS to something better, since it doesn't affect network protocol. I also still hate the way the function calculates netweight. I tried to adapt the old calculation and I just don't like it. I am going to gut it, rename it GetEstimatedNetworkWeight and simply have it call the new function GetAverageDifficulty and use the conversion factor I already worked out to convert the average difficulty to netweight.